Add filter by substation or VL to contingency list

flomillot opened the pull request:
- Introduced support for sub-equipment management in filter-based contingency lists.
- Added `FilterBasedContingencyListVisualizationPanel` for enhanced data visualization and management.
- Extended data schemas, form handling, and UI interactions to accommodate new sub-equipment feature.

To fix : 
- empty space at right of the Dialog
- responsive height on the whole Dialog, ie scroll bar on tables \ dialog when adding description or lot of filters...

flomillot committed:
Add filter by substation or VL to contingency list

- Introduced support for sub-equipment management in filter-based contingency lists.
- Added `FilterBasedContingencyListVisualizationPanel` for enhanced data visualization and management.
- Extended data schemas, form handling, and UI interactions to accommodate new sub-equipment feature.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

EtienneLt commented:
/home/lesoteti/Projects/Gridsuite/gridexplore-app/src/components/dialogs/contingency-list/filter-based/FilterBasedContingencyListVisualizationPanel.tsx
  22:18  warning  Insert `,`  prettier/prettier

flomillot committed:
Merge branch 'main' into filter-based-contingency-list

EtienneLt commented on src/translations/fr.json:
```diff
@@ -90,6 +90,8 @@
     "contingencyList.criteriaBasedOrExplicitNaming": "Par nommage ou par critères",
     "contingencyList.filterBased": "Par filtres",
+    "contingencyList.filterBased.filtersTableColumn": "Voltage level / substation filter",
+    "contingencyList.filterBased.subEquipmentsTableColumn": "Equipment types",
```
types d'equipement ?

EtienneLt commented on src/translations/fr.json:
```diff
@@ -90,6 +90,8 @@
     "contingencyList.explicitNaming": "Par nommage",
     "contingencyList.criteriaBasedOrExplicitNaming": "Par nommage ou par critères",
     "contingencyList.filterBased": "Par filtres",
+    "contingencyList.filterBased.filtersTableColumn": "Voltage level / substation filter",
```
filtres de postes / sites

EtienneLt reviewed (commented):
minor things

flomillot committed:
Update translations for contingency list table column labels in English and French

Signed-off-by: Florent MILLOT <florent.millot_externe@rte-france.com>

flomillot committed:
Replace `crypto` import with `node:crypto` and mark `VisualizationPanelProps` as `Readonly`

Signed-off-by: Florent MILLOT <florent.millot_externe@rte-france.com>

thangqp commented on src/utils/rest-api.ts:
```diff
@@ -575,15 +580,25 @@ export function getContingencyList(type: string, id: string) {
+            id: value.filterId,
+            equipmentTypes: value.subEquipmentTypes,
+        })),
+    };
```
Do not transform data in an API method; it is risky because you pass JSON.stringify(body) directly without type checking when you build the body object above

flomillot replied:
you re right i change it

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
using the same convention as others in explore-app for file/directory's name

flomillot replied:
done

flomillot committed:
Refactor contingency list filter-related functions to simplify parameters and type usage improvements

Signed-off-by: Florent MILLOT <florent.millot_externe@rte-france.com>

basseche commented on src/components/dialogs/contingency-list/filter-based/contingency-list-filter-based-dialog.tsx:
```diff
@@ -170,9 +191,13 @@ export default function FilterBasedContingencyListDialog({
-            unscrollableFullHeight
             disabledSave={Boolean(!!nameError || isValidating)}
             isDataFetching={isFetching}
+            sx={{
```
height are width are not correct. Height should be 90vh and width: 50vw.
if not possible keep unscrollableFullHeight and we'll try to fix it in another PR

flomillot replied:
As said by @thangqp  this morning, apparently it's pretty flexible, so I think I will stay like this

basseche commented on src/components/dialogs/contingency-list/filter-based/contingency-list-filter-based-form.tsx:
```diff
@@ -70,116 +81,109 @@ const equipmentColorsMap: Map<string, string> = new Map([
+    const [selectedFilterId, setSelectedFilterId] = useState<string | null>(null);
+    const [isDataOutdated, setIsDataOutdated] = useState<boolean>(false);
+
+    const filters: FilterElement[] = useWatch({ name: FieldConstants.FILTERS }) as FilterElement[];
```
add as unkown before FilterElement[] to turn off TypeScrip'st type assertion

flomillot replied:
Ah yes you're right

basseche commented on src/components/dialogs/contingency-list/filter-based/contingency-list-filter-based-form.tsx:
```diff
@@ -70,116 +81,109 @@ const equipmentColorsMap: Map<string, string> = new Map([
+    const substationAndVLFilters = filters.filter(isSubstationOrVoltageLevelFilter);
+    const filtersSubEquipments: FilterSubEquipments[] = useWatch({
+        name: ContingencyFieldConstants.SUB_EQUIPMENT_TYPES_BY_FILTER,
+    }) as FilterSubEquipments[];
```
same here

basseche commented on src/components/dialogs/contingency-list/filter-based/contingency-list-filter-based-form.tsx:
```diff
@@ -188,34 +192,90 @@ export default function ContingencyListFilterBasedForm() {
+                        </Grid>
+                        <Grid item xs={6}>
+                            <TableContainer component={Paper} sx={{ height: '100%', border: 0.5 }}>
+                                <Table>
```
I think you can reduce lineHeight by using size="small"

flomillot replied:
I'm not sure,  because the line height is the same for every row right now and if we make it smaller it's harder to click and it change the accessibility

thangqp replied:
IMO, it’s not critical! The table isn’t responsive vertically yet. If the upcoming fix doesn’t get done, we might need to use AG Grid.

flomillot committed:
Adapt contingency list filters to use `FiltersWithEquipmentTypes` type and fix casing for file import paths

Signed-off-by: Florent MILLOT <florent.millot_externe@rte-france.com>

flomillot committed:
Cast values returned by `useWatch` to `unknown` before applying specific types in contingency list filters

Signed-off-by: Florent MILLOT <florent.millot_externe@rte-france.com>

flomillot committed:
Rename `id` to `filterId` in `FilterBasedContingencyList` for clarity and consistency.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Improve clarity and consistency in `FilterBasedContingencyListVisualizationPanel`

- Renamed `id` to `filterId` in `selectedEquipmentTypesByFilter` for better readability.
- Adjusted grid layout properties and added padding for improved UI alignment.
- Refactored variable names and method calls for enhanced code clarity.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Rename `id` to `filterId` in `selectedEquipmentTypesByFilter` for consistency in `ContingencyListUtils`.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Adjust grid and dialog layout for improved UI responsiveness

- Added full-height styling to `Grid` in `ContingencyListFilterBasedForm`.
- Set dialog height to 95% of viewport in `ContingencyListFilterBasedDialog`.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Fix grid layout in `FilterBasedContingencyListVisualizationPanel`

- Adjusted `Grid` styling with a minimum width to address layout issues.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,214 @@
+    const shouldDisplayRefreshButton = selectedStudy?.length > 0 && isDataOutdated;
+
+    return (
+        <Grid item container direction="column" xs={3} sx={{minWidth: '31%'}}> {/* ugly width fix for the grid layout */}
```
```suggestion
        <Grid item container direction="column" xs> 
```

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,214 @@
+
+    return (
+        <Grid item container direction="column" xs={3} sx={{minWidth: '31%'}}> {/* ugly width fix for the grid layout */}
+            <Grid item xs={1}>
```
```suggestion
            <Grid item>
```

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,214 @@
+                    <FormattedMessage id="visualization" />
+                </Typography>
+            </Grid>
+            <Grid item container xs={2} paddingY={1} alignItems="center" justifyContent="center">
```
```suggestion
            <Grid item container paddingY={1} alignItems="center" justifyContent="center">
```

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,214 @@
+                </Grid>
+                <Grid item xs={7} fontWeight="bold" padding={1}>
+                    {selectedStudy.length > 0 ? (
+                        <Typography noWrap>
```
```suggestion
                        <Typography>
```

flomillot replied:
why? I don't think it's a good idea

thangqp commented on src/components/dialogs/contingency-list/filter-based/contingency-list-filter-based-dialog.tsx:
```diff
@@ -170,9 +191,14 @@ export default function FilterBasedContingencyListDialog({
             isDataFetching={isFetching}
+            sx={{
+                '.MuiDialog-paper': {
+                    minWidth: '60vw',
```
```suggestion
                    minWidth: '70vw',
```

flomillot replied:
it's already supposed to be fifty

thangqp replied:
OK for 60vw

flomillot committed:
Adjust grid container in `ContingencyListFilterBasedForm` for consistent layout

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Fix `shouldDisplayRefreshButton` logic in `FilterBasedContingencyListVisualizationPanel`

- Updated condition to include `isFetching` check.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Fix grid layout in `FilterBasedContingencyListVisualizationPanel`

- Removed redundant `xs` properties and adjusted `Grid` structure to simplify layout and maintain consistency.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Fix grid layout in `FilterBasedContingencyListVisualizationPanel`

- Removed redundant `xs` properties and adjusted `Grid` structure to simplify layout and maintain consistency.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Refactor overlay logic in `FilterBasedContingencyListVisualizationPanel`

- Consolidated `loadingOverlayComponent` and `loadingOverlayComponentParams` into a single `overlay` object.
- Simplified and clarified conditional logic for `shouldDisplayRefreshButton`.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Add warning alert for missing equipments in `FilterBasedContingencyListVisualizationPanel`

- Display a warning alert if study data has missing equipments.
- Update translations to include `missingEquipmentsFromStudy` entries in English and French.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Adjust grid margins in `ContingencyListFilterBasedForm` for improved alignment

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Fine-tune grid margins in `ContingencyListFilterBasedForm` for better alignment.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,223 @@
+    resizable: false,
+    sortable: false,
+};
+
```
```suggestion

export function RefreshOverlay({ loading, ...buttonProps }: Readonly<{ loading: boolean } & ButtonProps>) {
    return loading ? <CircularProgress /> : <RefreshButton {...buttonProps} />;
}

```

flomillot replied:
thank you for the tip, changed it

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,223 @@
+
+    const overlay = shouldDisplayRefreshButton
+        ? {
+              loadingOverlayComponent: RefreshButton,
```
```suggestion
              loadingOverlayComponent: RefreshOverlay,
```

flomillot replied:
done

thangqp commented on src/components/dialogs/contingency-list/filter-based/filter-based-contingency-list-visualization-panel.tsx:
```diff
@@ -0,0 +1,223 @@
+          }
+        : {
+              loadingOverlayComponentParams: { disabled: true }, // disable the button when loading
+              // rendering another loadingOverlayComponent or overlayLoadingTemplate does not work
```
```suggestion
              loadingOverlayComponentParams: { loading: true },
```

flomillot replied:
done

flomillot committed:
Enhance `FilterBasedContingencyListVisualizationPanel` functionality and refactor structure

- Added grid row focus capability with `focusOnRowById` using `AgGridReact`.
- Created `Overlay` component to handle loading and button rendering.
- Refactored row data handling by introducing `RowData` type and `getRowId` function.
- Improved `Typography` with better tooltips for folder and study display.
- Adjusted warning alert to trigger focus action on missing equipment rows.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Rename `focusOnRowById` to `scrollOnRowById` in `FilterBasedContingencyListVisualizationPanel` for clarity and update button usage accordingly.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Update button layout for missing equipment warning in `FilterBasedContingencyListVisualizationPanel`

- Encapsulated button within a `Grid` item for better alignment.
- Added `fullWidth` and `padding` styles to improve button presentation.

Signed-off-by: Florent MILLOT <millotflo@gmail.com>

flomillot committed:
Refactor separator row rendering in `FilterBasedContingencyListVisualizationPanel`

- Moved separator rendering logic to `fullWidthCellRenderer`.
- Simplified `cellRenderer` to handle non-separator rows only.
- Added padding for separator rows for better alignment.

Signed-off-by: Florent MILLOT <75525996+flomillot@users.noreply.github.com>

flomillot committed:
Update `@gridsuite/commons-ui` dependency to v0.132.0 in `package.json`

Signed-off-by: Florent MILLOT <75525996+flomillot@users.noreply.github.com>

flomillot committed:
Refactor and enhance UI components in `ContingencyListFilterBasedForm` and `FilterBasedContingencyListVisualizationPanel`

- Adjusted grid spacing for improved alignment in `ContingencyListFilterBasedForm`.
- Introduced `DataTableOverlay` component to replace inline `Overlay` logic for loading and refresh actions.
- Updated translations to include `equipmentTypesByFilters` and `reload` entries in English and French.
- Upgraded `@gridsuite/commons-ui` dependency to version `0.132.0`.

Signed-off-by: Florent MILLOT <75525996+flomillot@users.noreply.github.com>

flomillot committed:
Ensure separator row is scrolled into 'top' position in `FilterBasedContingencyListVisualizationPanel`.

Signed-off-by: Florent MILLOT <75525996+flomillot@users.noreply.github.com>

flomillot committed:
Handle nullish values for `name` and `equipmentType` in `contingency-list-utils`.

Signed-off-by: Florent MILLOT <75525996+flomillot@users.noreply.github.com>
